plugins/outputs/iotdb/iotdb.go

```diff
@@ -0,0 +1,287 @@
+	DataTypesList    [][]client.TSDataType
+	TimestampList    []int64
+	// extra tags
+	TagsList [][]*telegraf.Tag
+}
+
+func (*IoTDB) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init is for setup, and validating config.
+func (s *IoTDB) Init() error {
+	var errorMsg string
+	if s.Timeout < 0 {
+		errorMsg = fmt.Sprintf("IoTDB Config Warning: The value of 'timeout' is negative:%d. Now it's fixed to 0.", s.Timeout)
+		s.Log.Warnf(errorMsg)
+		s.Timeout = 0
+	}
+	if !(s.ConvertUint64To == "ToInt64" ||
+		s.ConvertUint64To == "ForceToInt64" ||
+		s.ConvertUint64To == "Text") {
+		errorMsg = fmt.Sprintf("IoTDB Config Warning: The value of 'ConvertUint64To' is invalid: %s. Now it's fixed to 'ToInt64'.", s.ConvertUint64To)
+		s.Log.Warnf(errorMsg)
+		s.ConvertUint64To = "ToInt64"
+	}
+	if !(s.TimeStampUnit == "second" ||
+		s.TimeStampUnit == "millisecond" ||
+		s.TimeStampUnit == "microsecond" ||
+		s.TimeStampUnit == "nanosecond") {
+		errorMsg = fmt.Sprintf("IoTDB Config Warning: The value of 'TimeStampUnit' is invalid: %s. Now it's fixed to 'nanosecond'.", s.TimeStampUnit)
+		s.Log.Warnf(errorMsg)
+		s.TimeStampUnit = "nanosecond"
+	}
+	if !(s.TreateTagsAs == "Measurements" || s.TreateTagsAs == "DeviceID_subtree") {
+		errorMsg = fmt.Sprintf("IoTDB Config Warning: The value of 'TreateTagsAs' is invalid: %s. Now it's fixed to 'Measurements'.", s.TreateTagsAs)
+		s.Log.Warnf(errorMsg)
+		s.TreateTagsAs = "Measurements"
+	}
+	s.Log.Info("IoTDB output plugin initialization completed.")
+	return nil
+}
+
+func (s *IoTDB) Connect() error {
+	// Make any connection required here
+	config := &client.Config{
+		Host:     s.Host,
+		Port:     s.Port,
+		UserName: s.User,
+		Password: s.Password,
+	}
+	var ss = client.NewSession(config)
+	s.session = &ss
+	if err := s.session.Open(false, s.Timeout); err != nil {
+		s.Log.Errorf("IoTDB Connect Error: Fail to connect host:'%s', port:'%s', err:%v", s.Host, s.Port, err)
+		return err
+	}
+
+	return nil
+}
+
+func (s *IoTDB) Close() error {
+	// Close any connections here.
+	// Write will not be called once Close is called, so there is no need to synchronize.
+	_, err := s.session.Close()
+	if err != nil {
+		s.Log.Errorf("IoTDB Close Error: %v", err)
+	}
+	return nil
+}
+
+// Write should write immediately to the output, and not buffer writes
+// (Telegraf manages the buffer for you). Returning an error will fail this
+// batch of writes and the entire batch will be retried automatically.
+func (s *IoTDB) Write(metrics []telegraf.Metric) error {
+	// Convert Metrics to Records with Tags
+	recordsWithTags, err := s.ConvertMetricsToRecordsWithTags(metrics)
+	if err != nil {
+		s.Log.Errorf(err.Error())
+		return err
+	}
+	// Wirte to client
+	// status, err := s.session.InsertRecords(deviceidList, measurementsList, dataTypesList, valuesList, timestampList)
+	err = s.WriteRecordsWithTags(recordsWithTags)
+	if err != nil {
+		s.Log.Errorf("IoTDB Write Error: %s", err.Error())
+	}
```

Comment:
Use local errors if possible
```suggestion
	if err := s.WriteRecordsWithTags(recordsWithTags); err != nil {
		s.Log.Errorf("IoTDB Write Error: %s", err.Error())
	}
```
Furthermore, why do you not error here? This way, the whole batch of data will be lost!

Reply:
Thanks for reminding! I will add a more sensible processing mechanism, such as rewriting when there is an error